listenbrainz/db/couchdb.py

```diff
@@ -0,0 +1,104 @@
+    """
+    databases_url = f"{get_base_url()}/{database}"
+    response = requests.put(databases_url)
+    response.raise_for_status()
+
+
+def list_databases(prefix: str) -> list[str]:
+    """ List all couchdb database whose name starts with the given prefix
+    sorted in the descending order of creation.
+
+    This method is useful because we name
+        Consider statistics, we generate those daily and create a new database each time
+    for each stat daily. After statistics for the day have been inserted, we want to
+    get rid of the older database for that stat. This method looks up all the databases
+    whose name starts with the stat name and then deletes all of those except the latest
+    one.
+    """
+    databases_url = f"{get_base_url()}/_all_dbs"
+    response = requests.get(databases_url)
+    response.raise_for_status()
+    all_databases = response.json()
+
+    databases = [database for database in all_databases if database.startswith(prefix)]
+    databases.sort(reverse=True)
+    return databases
+
+
+def delete_database(prefix: str):
+    """ Delete all but the latest database whose name starts with the given prefix.
+
+    Args:
+         prefix: the string to match database names with
+    """
+    databases = list_databases(prefix)
+    # remove the latest database from the list then delete the databases remaining in the list.
+    databases.pop(0)
+
+    for database in databases:
+        databases_url = f"{get_base_url()}/{database}"
+        response = requests.delete(databases_url)
+        response.raise_for_status()
+
+
+def fetch_data(prefix: str, user_id: int):
+    """ Retrieve data from couchdb for given stat type and user.
+
+    For each stat type, a database is created daily. We do not have a way to do this atomically so the latest
+    database for a type may be incomplete when we query it. So, query all databases for given stat 1 by 1 in
+    descending order of their creation until user data is found.
+
+    Args:
+         prefix: the string to match database names with
+         user_id: the user to retrieve data for
+    """
+    databases = list_databases(prefix)
+    base_url = get_base_url()
+
+    for database in databases:
+        document_url = f"{base_url}/{database}/{user_id}"
+        response = requests.get(document_url)
+        if response.status_code == 404:
+            continue
+        response.raise_for_status()
+        return response.json()
+
+    return None
+
+
+def insert_data(database: str, data):
+    """ Insert the given data into the specified database. """
+    with start_span(op="serializing", description="serialize data to json"):
+        docs = ujson.dumps({"docs": data})
+
+    with start_span(op="http", description="insert docs in couchdb using api"):
+        couchdb_url = f"{get_base_url()}/{database}/_bulk_docs"
+        response = requests.post(couchdb_url, data=docs, headers={"Content-Type": "application/json"})
+        response.raise_for_status()
```

Comment:
I wonder if we should be more intelligent about inserting the data here. If couchdb is unavailable, this will result in one sentry error message per user, which may not be all that good.

Reply:
The issue is that we don't persist state between messages so its either 1 error per user or none at all. However, sentry is often good at clustering messages. For instance:

![image](https://user-images.githubusercontent.com/27751938/178478031-8157a582-5d6d-416f-8718-ec07f3da42c5.png)

So, it does have 22k error reports but those are grouped together so we can view it together as one.